platform/packages/platform/src/reply.rs

```diff
@@ -23,78 +25,85 @@ impl InstantiateResponse<Vec<u8>> {
     from_instantiate_inner::<MsgInstantiateContractResponse>(api, reply)
         .map(InstantiateResponse::into_addr)
 }
 
-pub fn from_instantiate2_addr_only(api: &dyn Api, reply: Reply) -> StdResult<Addr> {
+pub fn from_instantiate2_addr_only(api: &dyn Api, reply: Reply) -> Result<Addr> {
     from_instantiate_inner::<MsgInstantiateContract2Response>(api, reply)
         .map(InstantiateResponse::into_addr)
 }
 
-pub fn from_execute<T>(reply: Reply) -> StdResult<Option<T>>
+pub fn from_execute<T>(reply: Reply) -> Result<Option<T>>
 where
     T: DeserializeOwned,
 {
     decode::<MsgExecuteContractResponse>(reply)
         .map(|data| data.data)
-        .map(Binary)
-        .and_then(from_json)
+        .and_then(|data| from_json(Binary(data)).map_err(From::from))
 }
 
-struct UncheckedInstantiateResponse {
-    address: String,
-    data: Vec<u8>,
+trait InstantiationResponse
+where
+    Self: Message + Default + Sized,
+{
+    fn addr(&self) -> &str;
+
+    fn into_data(self) -> Vec<u8>;
 }
 
-impl From<MsgInstantiateContractResponse> for UncheckedInstantiateResponse {
-    fn from(
-        MsgInstantiateContractResponse { address, data }: MsgInstantiateContractResponse,
-    ) -> Self {
-        Self { address, data }
+impl InstantiationResponse for MsgInstantiateContractResponse {
+    fn addr(&self) -> &str {
+        &self.address
+    }
+
+    fn into_data(self) -> Vec<u8> {
+        self.data
     }
 }
 
-impl From<MsgInstantiateContract2Response> for UncheckedInstantiateResponse {
-    fn from(
-        MsgInstantiateContract2Response { address, data }: MsgInstantiateContract2Response,
-    ) -> Self {
-        Self { address, data }
+impl InstantiationResponse for MsgInstantiateContract2Response {
+    fn addr(&self) -> &str {
+        &self.address
+    }
+
+    fn into_data(self) -> Vec<u8> {
+        self.data
     }
 }
 
-fn from_instantiate_inner<R>(api: &dyn Api, reply: Reply) -> StdResult<InstantiateResponse<Vec<u8>>>
+fn from_instantiate_inner<R>(api: &dyn Api, reply: Reply) -> Result<InstantiateResponse<Vec<u8>>>
 where
-    R: Message + Default + Into<UncheckedInstantiateResponse>,
+    R: InstantiationResponse,
 {
-    let UncheckedInstantiateResponse { address, data } = decode::<R>(reply)?.into();
+    let response: R = decode(reply)?;
 
-    api.addr_validate(&address)
-        .map(|address: Addr| InstantiateResponse { address, data })
+    api.addr_validate(response.addr())
+        .map(|address: Addr| InstantiateResponse {
+            address,
+            data: response.into_data(),
+        })
+        .map_err(From::from)
 }
 
-fn decode_raw<M>(message: &[u8]) -> StdResult<M>
+fn decode_raw<M>(message: &[u8]) -> Result<M>
 where
     M: Message + Default,
 {
-    M::decode(message).map_err(|error| {
-        StdError::generic_err(format!("[Platform] Data is malformed or doesn't comply with used protobuf format! Cause: [Protobuf] {error}"))
-    })
+    M::decode(message).map_err(From::from)
```

Comment:
Is the usage of `From` instead of `Into` intentional? As far as I remember the books say "define From, use Into"

Reply:
Yes, it does say it's preferred. My train of thought was that it might introduce a second, nested, call but on closer inspection the implementation is indeed marked with `#[inline]`. The more you know. Will make it use `Into` right away.

Reply:
I think that it meant to prefer using `Into` when generics are involved, as that is where it might cause a problem.